File: app/src/main/java/com/pr0gramm/app/services/SingleShotService.java
package com.pr0gramm.app.services;

import android.content.SharedPreferences;

import com.google.gson.Gson;
import com.pr0gramm.app.util.AndroidUtility;

import org.joda.time.DateTime;
import org.joda.time.format.DateTimeFormat;

import java.util.Collections;
import java.util.HashMap;
import java.util.HashSet;
import java.util.Map;
import java.util.Set;

import javax.inject.Inject;
import javax.inject.Singleton;

/**
 */
@Singleton
public class SingleShotService {
    static final int TIME_OFFSET_IN_MILLIS = (int) (Math.random() * 3600 * 1000);

    private static final String KEY_ACTIONS = "SingleShotService.actions";
    private static final String KEY_MAP_ACTIONS = "SingleShotService.mapActions";

    private final Gson gson = new Gson();
    final SharedPreferences preferences;

    Map<String, String> timeStringMap;

    final Object lock = new Object();


    @Inject
    public SingleShotService(SharedPreferences preferences) {
        this.preferences = preferences;
    }

    public synchronized boolean isFirstTime(String action) {
        synchronized (lock) {
            Set<String> actions = new HashSet<>(preferences.getStringSet(
                    KEY_ACTIONS, Collections.<String>emptySet()));

            if (actions.add(action)) {
                // store modifications
                preferences.edit().putStringSet(KEY_ACTIONS, actions).apply();
                return true;
            } else {
                return false;
            }
        }
    }

    public boolean firstTimeInVersion(String action) {
        int version = AndroidUtility.buildVersionCode();
        return isFirstTime(action + "--" + version);
    }

    public boolean firstTimeToday(String action) {
        return firstTimeByTimePattern(action, "YYYY-MM-dd");
    }

    public boolean firstTimeInHour(String action) {
        return firstTimeByTimePattern(action, "YYYY-MM-dd:HH");
    }

    public boolean firstTimeByTimePattern(String action, String pattern) {
        String timeString = DateTime.now()
                .minusMillis(TIME_OFFSET_IN_MILLIS)
                .toString(DateTimeFormat.forPattern(pattern));

        return timeStringHasChanged(action, timeString);
    }

    public TestOnlySingleShotService test() {
        return new TestOnlySingleShotService();
    }

    public class TestOnlySingleShotService {
        TestOnlySingleShotService() {
        }

        public boolean isFirstTime(String action) {
            synchronized (lock) {
                return !preferences
                        .getStringSet(KEY_ACTIONS, Collections.<String>emptySet())
                        .contains(action);
            }
        }

        public boolean firstTimeInVersion(String action) {
            int version = AndroidUtility.buildVersionCode();
            return isFirstTime(action + "--" + version);
        }

        public boolean firstTimeToday(String action) {
            return firstTimeByTimePattern(action, "YYYY-MM-dd");
        }

        public boolean firstTimeInHour(String action) {
            return firstTimeByTimePattern(action, "YYYY-MM-dd:HH");
        }

        public boolean firstTimeByTimePattern(String action, String pattern) {
            String timeString = DateTime.now()
                    .minusMillis(TIME_OFFSET_IN_MILLIS)
                    .toString(DateTimeFormat.forPattern(pattern));
            return timeStringHasChanged(action, timeString);
        }

        private boolean timeStringHasChanged(String action, String timeString) {
            synchronized (lock) {
                return timeStringMap == null || !timeString.equals(timeStringMap.get(action));
            }
        }
    }

    @SuppressWarnings("unchecked")
    private boolean timeStringHasChanged(String action, String timeString) {
        synchronized (lock) {
            if (timeStringMap == null) {
                try {
                    timeStringMap = new HashMap<>(gson.fromJson(
                            preferences.getString(KEY_MAP_ACTIONS, "{}"),
                            Map.class));

                } catch (RuntimeException ignored) {
                    timeStringMap = new HashMap<>();
                }
            }

            if (timeString.equals(timeStringMap.get(action))) {
                return false;
            } else {
                timeStringMap.put(action, timeString);

                preferences.edit()
                        .putString(KEY_MAP_ACTIONS, gson.toJson(timeStringMap))
                        .apply();

                return true;
            }
        }
    }
}

File: app/src/main/java/com/pr0gramm/app/services/SingleShotService.kt
@file:Suppress("PLATFORM_CLASS_MAPPED_TO_KOTLIN")

package com.pr0gramm.app.services

import android.content.SharedPreferences
import androidx.core.content.edit
import com.pr0gramm.app.Duration
import com.pr0gramm.app.Instant
import com.pr0gramm.app.MoshiInstance
import com.pr0gramm.app.util.AndroidUtility
import com.pr0gramm.app.util.getStringOrNull
import com.squareup.moshi.adapter
import java.text.SimpleDateFormat
import java.util.*

class SingleShotService(private val preferences: SharedPreferences) {
    private val timeOffset = Duration.millis((Math.random() * 3600.0 * 1000.0).toLong())

    private val keySetActions = "SingleShotService.actions"
    private val keyMapActions = "SingleShotService.mapActions"

    @Suppress("PLATFORM_CLASS_MAPPED_TO_KOTLIN")
    private val mapAdapter = MoshiInstance.adapter<java.util.Map<String, String>>()

    private val lock = Any()

    private val timeStringMap: MutableMap<String, String> = loadTimeStringMap()

    /**
     * Marks the given action as done and returns true, if it was not yet done.
     */
    fun markAsDoneOnce(action: String): Boolean {
        synchronized(lock) {
            val actions = (preferences.getStringSet(keySetActions, null) ?: setOf()).toMutableSet()

            return if (actions.add(action)) {
                preferences.edit {
                    putStringSet(keySetActions, actions)
                }

                true
            } else {
                false
            }
        }
    }

    fun isFirstTime(action: String): Boolean {
        synchronized(lock) {
            val actions = preferences.getStringSet(keySetActions, null) ?: setOf()
            return action !in actions
        }
    }

    inline fun doOnce(action: String?, block: () -> Unit) {
        if (action == null) {
            block()
        } else if (markAsDoneOnce(action)) {
            block()
        }
    }

    fun firstTimeInVersion(action: String): Boolean {
        val version = AndroidUtility.buildVersionCode()
        return markAsDoneOnce("$action--$version")
    }

    fun firstTimeToday(action: String): Boolean {
        return firstTimeByTimePattern(action, "yyyy-MM-dd")
    }

    fun firstTimeInHour(action: String): Boolean {
        return firstTimeByTimePattern(action, "yyyy-MM-dd:HH")
    }

    private fun firstTimeByTimePattern(action: String, pattern: String): Boolean {
        return timeStringHasChanged(action, Instant.now()
                .minus(timeOffset)
                .toString(SimpleDateFormat(pattern, Locale.ROOT)))
    }

    private fun timeStringHasChanged(action: String, timeString: String): Boolean {
        synchronized(lock) {
            return if (timeStringMap[action] == timeString) {
                false
            } else {
                timeStringMap[action] = timeString

                preferences.edit {
                    @Suppress("PLATFORM_CLASS_MAPPED_TO_KOTLIN", "UNCHECKED_CAST")
                    putString(keyMapActions, mapAdapter.toJson(HashMap<String, String>(
                            timeStringMap) as java.util.Map<String, String>))
                }

                true
            }
        }
    }

    @Suppress("UNCHECKED_CAST")
    private fun loadTimeStringMap(): MutableMap<String, String> {
        try {
            val encoded = preferences.getStringOrNull(keyMapActions) ?: "{}"
            val map = mapAdapter.fromJson(encoded)
            if (map != null) {
                val kMap = map as Map<String, String>
                return kMap.toMutableMap()
            }

        } catch (ignored: RuntimeException) {
        }

        return HashMap()
    }
}
